Adding support for deleteFilesInRanges in JNI

koldat opened the pull request:
It is very useful method call to achieve https://github.com/facebook/rocksdb/wiki/Delete-A-Range-Of-Keys

adamretter commented on java/src/test/java/org/rocksdb/RocksDBTest.java:
```diff
@@ -602,6 +602,60 @@ public void compactRangeToLevel()
+        setWriteBufferSize(FILE_SIZE / 2).
+        setNumLevels(5).
+        setDisableAutoCompactions(true);
+         final RocksDB db = RocksDB.open(opt,
```
I think several of these options and unnecessary for the test and could be removed, to make the intention clearer.

koldat commented on java/rocksjni/rocksjni.cc:
```diff
@@ -2254,3 +2255,73 @@ void Java_org_rocksdb_RocksDB_destroyDB(JNIEnv* env, jclass /*jcls*/,
+  std::vector<rocksdb::RangePtr> rangesVector;
+  std::vector<std::unique_ptr<rocksdb::Slice>> slices;
+  std::vector<std::unique_ptr<jbyte[]>> ranges_to_free;
+  for (jsize i = 0; (i + 1) < length; i += 2) {
```
The main point is that I did not want to care about allocations like on other places in this file. I am relying on unique_ptr so that any exception (including C++ one) will not create a leak. So ranges_to_free are freed by C++ destructor of vector -> then unique_ptr -> delete jbyte[].

adamretter replied:
I see.
p.s. you do also allocate in `get_slice_helper` already though for the `jbyte[]`

koldat replied:
Intention is this. I wanted to reuse the algorithm for "sliceFrom" and "sliceTo" and without any possibility for a leak.
rangesVector - this is holding RangePtr instances that I can reference as pointer and size do deleteFiles
slices - this is holding auto pointer to Slice. RangePtr is referencing Slice* so I need to hold it somewhere with safe free and that is this vector
ranges_to_free - this is holding any dynamically allocated jbyte[] It is referenced in slice as pointer so I need to hold it somewhere with safe free and that is this vector

get_slice_helper puts jbyte[] into our safe container and then sets pointer to it in slice.

If anything happen, everything will be freed without leak. One cannot forget to call something like multi_get_helper_release_keys.

adamretter commented on java/src/test/java/org/rocksdb/RocksDBTest.java:
```diff
@@ -602,6 +602,60 @@ public void compactRangeToLevel()
+      int records = FILE_SIZE / (KEY_SIZE + VALUE_SIZE);
+
+      // fill database with key/value pairs
+      byte[] value = new byte[VALUE_SIZE];
```
Could we show the same results with less key value pairs and less files? Just thinking about execution time...

koldat replied:
Execution time is 1 second. I was trying to make files smaller, but compactions are not predicable much. This combination is stable. I will try to squeeze that little more.

adamretter replied:
1 second is fine. I thought from the options that auto compactions were disabled? Maybe I am not understanding something, I have not worked a lot with testing around compactions

adamretter commented on java/src/test/java/org/rocksdb/RocksDBTest.java:
```diff
@@ -602,6 +602,60 @@ public void compactRangeToLevel()
+
+      // Should be 10
+      int files = Integer.parseInt(db.getProperty("rocksdb.num-files-at-level1"));
+      assertThat(files).isBetween(8, 12);
```
The comment states this should be `10` but you are asserting `8 to 12`?

koldat replied:
Yes. I am not asserting exact numbers, but that delete files actually deletes something. I did not want to introduce unstable test.

adamretter replied:
So, would the instability would come due to the "compactions are not predicable" issue you mentioned?

koldat replied:
Yes. Sometimes even if you specify targetSize it is using memory log on level 0 and ignoring target size. Actually compaction then just promotes file to higher level, because there is no overlap thus size is bigger than expected for test. That is why I am doing that crazy interleaved puts to force worst scenario so that compaction end up with "nice" files I can delete. Because we are not trying to assert logic in C++ I am fine with these acceptable ranges.

adamretter commented on java/src/test/java/org/rocksdb/RocksDBTest.java:
```diff
@@ -602,6 +602,60 @@ public void compactRangeToLevel()
+      // Delete half (roughly)
+      db.deleteFilesInRanges(null, Arrays.asList(null, String.format("%020d", records * KEY_INTERVAL / 2).getBytes()), false);
+      files = Integer.parseInt(db.getProperty("rocksdb.num-files-at-level1"));
+      assertThat(files).isBetween(3, 7);
```
Again, is it possible to be more accurate than using a range?

koldat replied:
It can cause unstability. Intention is to test JNI and not actual RocksDB. It has its own detailed unit test for this functionality.

adamretter replied:
Understood :-)

koldat commented on java/src/main/java/org/rocksdb/RocksDB.java:
```diff
@@ -2174,6 +2174,30 @@ public void ingestExternalFile(final ColumnFamilyHandle columnFamilyHandle,
+   * @throws RocksDBException thrown if error happens in underlying
+   *     native library.
+   */
+  public void deleteFilesInRanges(ColumnFamilyHandle column_family, final List<byte[]> ranges, boolean include_end) throws RocksDBException {
```
OK. I have copied that from compactRange where is same issue. I will rename it.

adamretter replied:
Thanks :-)

adamretter commented:
@koldat thanks for your contribution in general it looks good. I have a few minor comments/suggestions

koldat commented:
@adamretter I have pushed changeset with changes from your review. It is squashed to be ready for merge. What I did:
- camel casing
- Lowered size of test, removed unnecessary options and added comment of what I am trying to achieve so it will be easier to understand. Runtime is around 800ms on my box.

Can you confirm you are ok?

koldat commented:
@adamretter Are you OK with my latest changes based on your review?

christian-esken commented:
I would love to see this deleteFilesinRange PR merged, as I have a use case for it. Just wanted to chip in, as I am curious about the status of this PR.

adamretter commented:
@ajkr @sagar0 this one can be merged please.

siying commented:
@koldat we are sorry that we forgot to merge it. Can you help do rebase and then we'll merge it.

koldat commented:
I have rebased the change

vjnadimpalli commented:
@koldat Thanks a lot for this contribution. Once all the tests pass we should be able to merge it. There seems to be some problem with the appveyor tests would you mind taking a look?

koldat commented:
It was unrelated issue fixed yesterday.  #5788

vjnadimpalli commented:
@koldat maybe try to get that change and re-run appveyor tests to make sure it passes. Also 'make format' so linter doesn't complain. Trying all greens before landing.

koldat commented:
How to rerun test? I don't know how to do that.

vjnadimpalli commented:
For Travis within the job there are restart options, for Appveyor I don't see that. I think if you just make format and commit that change, it should re-run the tests. Hopefully you can rebase on top of #5788.

koldat commented:
I have made "make format" and squashed rebased to current master
